firebase-firestore/src/main/java/com/google/firebase/firestore/core/UserData.java

```diff
@@ -303,9 +301,15 @@ private void validatePath() {
       }
     }
 
+    private static final String RESERVED = "__";
+
     private void validatePathSegment(String segment) {
-      if (isWrite() && reservedFieldRegex.matcher(segment).find()) {
-        throw this.createError("Document fields cannot begin and end with __");
+      if (segment.isEmpty()) {
+        throw this.createError("Document fields must not be empty");
+      }
+
+      if (isWrite() && segment.startsWith(RESERVED) && segment.endsWith(RESERVED)) {
+        throw this.createError("Document fields cannot begin or end with \"__\"");
```

Comment:
I think the original error message ("begin and end with") is more precise, no?

Reply:
Yes. This was part of my think-o. Thanks for persisting :-).